Source/Layout/ASLayoutElementContext.mm

```diff
@@ -0,0 +1,124 @@
+    _transitionID = ASLayoutElementContextDefaultTransitionID;
+    _pendingLayoutMap = std::make_shared<std::unordered_map<void *, std::shared_ptr<ASDisplayNodeLayout>>>();
+  }
+  return self;
+}
+
+@end
+
+CGFloat const ASLayoutElementParentDimensionUndefined = NAN;
+CGSize const ASLayoutElementParentSizeUndefined = {ASLayoutElementParentDimensionUndefined, ASLayoutElementParentDimensionUndefined};
+
+int32_t const ASLayoutElementContextInvalidTransitionID = 0;
+int32_t const ASLayoutElementContextDefaultTransitionID = ASLayoutElementContextInvalidTransitionID + 1;
+
+pthread_key_t ASLayoutElementContextKey;
+
+static void ASLayoutElementDestructor(void *p) {
+  if (p != NULL) {
+    ASDisplayNodeCFailAssert(@"Thread exited without clearing layout element context!");
+    CFBridgingRelease(p);
+  }
+};
+
+// pthread_key_create must be called before the key can be used. This function does that.
+void ASLayoutElementContextEnsureKey()
+{
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    pthread_key_create(&ASLayoutElementContextKey, ASLayoutElementDestructor);
+  });
+}
+
+void ASLayoutElementPushContext(ASLayoutElementContext *context)
+{
+  ASLayoutElementContextEnsureKey();
+  // NOTE: It would be easy to support nested contexts – just use an NSMutableArray here.
+  ASDisplayNodeCAssertNil(ASLayoutElementGetCurrentContext(), @"Nested ASLayoutElementContexts aren't supported.");
+  pthread_setspecific(ASLayoutElementContextKey, CFBridgingRetain(context));
+}
+
+ASLayoutElementContext *ASLayoutElementGetCurrentContext()
+{
+  ASLayoutElementContextEnsureKey();
+  // Don't retain here. Caller will retain if it wants to!
+  return (__bridge __unsafe_unretained ASLayoutElementContext *)pthread_getspecific(ASLayoutElementContextKey);
+}
+
+void ASLayoutElementPopContext()
+{
+  ASLayoutElementContextEnsureKey();
+  ASDisplayNodeCAssertNotNil(ASLayoutElementGetCurrentContext(), @"Attempt to pop context when there wasn't a context!");
+  CFBridgingRelease(pthread_getspecific(ASLayoutElementContextKey));
+  pthread_setspecific(ASLayoutElementContextKey, NULL);
+}
+
+std::shared_ptr<ASDisplayNodeLayout> ASLayoutElementContextGetPendingLayout(ASDisplayNode *node)
+{
+  if (node == nil) {
+    return nullptr;
+  }
+  
+  void *nodePointer = (__bridge void *)node;
+  std::shared_ptr<std::unordered_map<void *,
+                  std::shared_ptr<ASDisplayNodeLayout>>> pendingLayoutMap =
+                                                            ASLayoutElementGetCurrentContext().pendingLayoutMap;
+  if (pendingLayoutMap == nullptr) {
+    // WARNING: This condition should probably not be reached. It implies that the
+    // caller is asking for _pendingLayout outside of any layout operation.
+    return nullptr;
+  }
+
+  return (*(ASLayoutElementGetCurrentContext().pendingLayoutMap))[nodePointer];
+}
+
+void ASLayoutElementContextSetPendingLayout(ASDisplayNode *node, std::shared_ptr<ASDisplayNodeLayout> layout)
```

Comment:
Ah, we're gonna store the whole `ASDisplayNodeLayout` instead of just `ASLayout`. Clever!